File: algorithms/sorting/Java/MergeSort.java
import java.util.Arrays;
import java.util.Random;

/*
 * Time Complexity:
 *   Omega(nlogn)
 *   Theta(nlogn)
 *   O(nlogn)
 *
 * Space Complexity:
 *   O(n)
 */
class MergeSort {

    public static void main(String[] args) {
        
        Random rand = new Random();
        int[] nums = new int[20];

        for (int i = 0; i < nums.length; ++i) {
            nums[i] = rand.nextInt(101);
        }

        System.out.println("Unsorted: " + Arrays.toString(nums));
        sort(nums, 0, nums.length-1);
        System.out.println("Sorted: " + Arrays.toString(nums));
    }

    public static void sort(int[] arr, int p, int r) {
        if (p < r) {
            int q = (p + r) / 2;
            sort(arr, p, q);
            sort(arr, q + 1, r);
            merge(arr, p, q, r);
        }
    }

    public static void merge(int[] arr, int p, int q, int r) {
        int[] lo = new int[q - p + 1];
        int[] hi = new int[r - q];
        int i, j, k;

        k = p;
        i = j = 0;
        while (k <= q) lo[i++] = arr[k++];
        while (k <= r) hi[j++] = arr[k++];

        k = p;
        i = j = 0;
        while (i < lo.length && j < hi.length) {
            if (lo[i] < hi[j]) arr[k++] = lo[i++];
            else arr[k++] = hi[j++];
        }

        while (i < lo.length) arr[k++] = lo[i++];
        while (j < hi.length) arr[k++] = hi[j++];
    }
}

File: algorithms/sorting/Python3/mergesort.py
#
# Merge Sort
#

import random

def mergesort(list_):
    # Divide the list
    if len(list_) > 1:
        middle = len(list_) // 2
        left = list_[:middle]
        right = list_[middle:]

        # Divide recursively
        mergesort(left)
        mergesort(right)

        i = 0 # Left iterator
        j = 0 # Right iterator
        k = 0 # List iterator

        # Merge
        while i < len(left) and j < len(right):
            if left[i] < right[j]:
                list_[k] = left[i]
                i += 1
            else:
                list_[k] = right[j]
                j += 1
            k += 1

        # Empty remaining values from left half
        while i < len(left):
            list_[k] = left[i]
            i += 1
            k += 1

        # Empty remaining values from right half
        while j < len(right):
            list_[k] = right[j]
            j += 1
            k += 1

def main():
    list_ = random.sample(range(101), 10)
    print("Unsorted:", list_)
    mergesort(list_)
    print("Sorted:  ", list_)


if __name__ == "__main__":
    main()
